[그리디] 이창희 자동차 경주미션 1,2단계 미션 제출합니다.

chxghee opened the pull request:
안녕하세요 리뷰어님 
이번에 그리디 2기로 참여하게 된 이창희라고 합니다~
앞으로 잘 부탁드립니다!

제가 이번 미션을 진행하면서 가장 고민했던 점은 크게 

1. 랜덤값을 생성하여 사용하는 메서드의 테스트를 어떻게 진행할 것인가?
2. 제가 만든 레이싱이라는 클래스에 우승한 자동차들의 리스트를 필드로 저장하여 유지할 필요가 있을까?

일단 제가 나름 고민한 결과

1번 -> 최대한 메서드를 분리하여 난수값에 영향을 받지 않는 메서드라도 테스트를 하자.
2번 -> 레이싱이 끝난 후 레이싱의 우승자를 빈번하게 조회하지 경우가 아니라면, 굳이 필드로 저장해 놓지 않고 메서드를 통해 우승자 리스트를 반환하도록 만드는 것이 나은 것 같다.  

라는 결론을 내려 과제를 제출하였습니다.

제가 고민하고 나름 결론을 내렸던 부분에 대한 리뷰어님의 생각이 궁금합니다!

chxghee committed:
Docs: 1,2단계 README 작성

1,2 단계 요구사항 정리 및 구현 방향성에 대한 전략 작성

chxghee committed:
Update: README 수정

줄바꿈 수정

chxghee committed:
Docs: README 줄바꿈 수정

chxghee committed:
Feat: 자동차 이동 및 우승자 집계 기능 구현

chxghee committed:
Fix: 난수 생성 범위 수정

chxghee committed:
Refactor: 패키지명 변경 (entity -> domain)

be-student commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    public Boolean isWinner(int winningPosition) {
+        if (this.position == winningPosition) {
+            return true;
+        }
+        return false;
+    }
```
이 메소드 시그니처가 직관적이지 않은데요
이 메소드가 Car 클래스에 있어야 할까? 하는 생각이 드는 것 같아요

제가 생각하는 객체지향을 사용하는 가장 큰 장점은 특정 클래스 or 메소드만 보고서도 코드의 흐름을 파악하기 쉬운 것이 큰 장점인 것 같은데요

이때 이 Car 클래스의 메소드인 isWinner 라는 것은 약간 애매한 것 같아요
아래 방식 2가지중 하나로 바꿔보면 어떻게 생각하시는지 궁금합니다

1.  저라면 Racing 클래스에서 .filter(car -> car.getPosition()==winningPosition) 와 같은 형태로 진행할 것 같아요


2번은 suggestion 으로 작성해두었습니다

```suggestion
    public boolean hasSamePosition(int position) {
        if (this.position == position) {
            return true;
        }
        return false;
    }
```
코드를 읽는 입장에서 isWinner 로 되어있는 코드를 읽는 편과 단순 getter 를 쓰거나 hasSamePosition 같은 네이밍이면 조금 더 인지에 들어가는 피로도를 줄일 수 있을 것 같아요

저는 메소드 네이밍이나 가독성을 크게 신경쓰는 편은 아니라 의견이 다를 수 있어요
제 의견들은 어떻게 생각하시는지 달아주시면 감사하겠습니다!
마음에 든다면 수정해주셔도 좋고, 아니라면 유지해주셔도 좋아요

chxghee replied:
지금 생각해 보니  .filter(car -> car.getPosition()==winningPosition) 로 처리해도 가독성있게 읽을 수 있는 코드를 불필요하게 isWinning 이라는 메서드를 만든 것 같다고 생각이 드네요

isWinning이라는 메서드의 네이밍이 너무 Racing 클래스의 비즈니스에 특화된 네이밍인것 같고, 게다가 이 메서드가 Car 클래스에 위치하니 더욱 이해하기 힘드셨을 것 같습니다

또한 isWinning 메서드가 Racing 클래스 내에 위치해 있다고 하더라도, isWinning라고 네이밍을 하게 되면 해당 메서드를 재사용하기 어려워 지는 것 같네요

말씀해 주신대로 수정을 하는게 맞는것 같습니다! 감사합니다

be-student commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    public Boolean isWinner(int winningPosition) {
```
Boolean 대신 boolean 을 고려해보시면 더 좋을 것 같아요!
아마 아시겠지만 Boolean 은 wrapper 타입이라서 그만큼 메모리가 더 사용되는 것과 npe 가 발생할 수 있다는 점에서 안좋을 수 있을 것 같아요

be-student commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,28 @@
+        return position;
+    }
+
+    public void moveCar(int random) {
+        if (random >= 4) {
+            this.position += 1;
+        }
+    }
```
여기의 4는 상수로 분리해보면 어떨까요?
1은 뭐 1씩 움직이는게 되게 직관적으로 보이지만 4 이상이어야 한다는 것은 일반적인 코드에서 많이 사용되는 숫자가 아니다보니까 상수나 주석으로 설명을 달아주면 좋을 것 같아요

be-student commented on src/main/java/domain/Racing.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void round() {
+        for (Car car : cars) {
+            Random random = new Random();
```
Random 자체도 성능상 좋지 않을 수 있고, 매번 객체를 만드는 것은 성능상 좋지 않은데요
다양한 Random 에 대해서는
https://morethantoday.tistory.com/94
이 글을 통해서 학습해보셔도 좋을 것 같아요!

Random 은 상수로 분리하거나 or ThreadLocalRandom 과 같은 다른 방식을 통해서 생성해주시면 더 좋을 것 같습니다

be-student commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,31 @@
+
+        Assertions.assertThat(car.isWinner(winningPosition)).isTrue();
+    }
+}
```
깃허브에서 파일을 보면 
`-` 표시가 파일 끝에 남아있게 되는데요
https://jojoldu.tistory.com/673
이 글을 참고하셔서 intellij 의 설정을 해보셔도 좋을 것 같아요

be-student commented on src/test/java/domain/RacingTest.java:
```diff
@@ -0,0 +1,46 @@
+        racing.createCars(carNum);
+    }
+
+    @Test
+    void 자동차는_입력한_개수만큼_리스트에_추가된() {
+        Assertions.assertThat(racing.getCars().size()).isEqualTo(5);
+    }
```
```suggestion
    @Test
    void 자동차는_입력한_개수만큼_리스트에_추가된다() {
        Assertions.assertThat(racing.getCars().size()).isEqualTo(5);
    }
```

be-student commented on src/test/java/domain/RacingTest.java:
```diff
@@ -0,0 +1,46 @@
+        Assertions.assertThat(racing.getCars().size()).isEqualTo(5);
+    }
+
+    @Test
+    void 가장_먼_위치의_있는_자동차의_위치가_우승_위치이다() {
+        List<Car> cars = racing.getCars();
+        int i = 0;
+        for (Car car : cars) {
+            car.moveCar(i++);
+        }
+        Assertions.assertThat(racing.findWinningPosition()).isEqualTo(1);
+    }
```
테스트코드에서 i=0 부터 i++ 를 하는 등의 로직이 들어가게 되면 테스트코드를 테스트하는 테스트코드가 필요해지는 날이 생길 수 있을 것 같아요
테스트코드의 복잡도를 낮춰보면 어떨까요?

be-student reviewed (changes requested):
안녕하세요 창희님 미션 하시느라 고생 많으셨습니다

정말 잘 해주셔서 조금 더 빡빡하게 코드를 보게 되었네요

1. 랜덤값을 생성하여 사용하는 메서드의 테스트를 어떻게 진행할 것인가?
> 1번 -> 최대한 메서드를 분리하여 난수값에 영향을 받지 않는 메서드라도 테스트를 하자.

제가 생각하는 좋은 메소드는 테스트하기 좋고, 이해하기 좋은 구조를 가진 코드입니다
먼저 테스트하기 좋은 메소드의 경우에는 
https://jojoldu.tistory.com/674 
이 글에서 되게 잘 나와있는 것 같아서 참고로 달아드려요

저는 개인적으로 Random 값은 이것처럼 테스트를 해야 한다고 생각하고 있어요
https://xxeol.tistory.com/11
마침 비슷한 고민을 하시는 것 같아서 참고로 달아드려요

이렇게 1차적으로 분리가 되고 결과적으로는 설정을 통해서 바꿀 수 있도록 관리되도록 하는 것이 가장 좋은 방향인 것 같아요 (테스트 환경이라면 원하는 결과를 받을 수 있도록 주입해준다던가?)


-----------------

2. 제가 만든 레이싱이라는 클래스에 우승한 자동차들의 리스트를 필드로 저장하여 유지할 필요가 있을까?
> 레이싱이 끝난 후 레이싱의 우승자를 빈번하게 조회하지 경우가 아니라면, 굳이 필드로 저장해 놓지 않고 메서드를 통해 우승자 리스트를 반환하도록 만드는 것이 나은 것 같다.

일반적인 경우에 성능을 고려해서 작성한다고하면 필드로 저장해두는 것을 무조건적으로 권장드리는 편인데요
정확하게는 startRacing 의 반환값이 레이싱 결과를 담은 객체인 경우가 많을 것 같고요
결과를 저장해서 무조건 O(1) 로 조회가 가능하도록 하는 것이 도움이 되는 것 같습니다

저는 아무리 간단한 결과라고 해도 보통 이렇게 두는 것을 선호하는데요
공수가 그렇게까지 크게 들어가지 않는 것 + 추후에 결과를 다른 곳에서 활용하게 되면 그때마다 매번 계산하도록 하는 것이 한정된 서버 리소스를 비효율적으로 사용하게 되다보니 그런 것 같습니다

----------

궁금하신 것들이 있으면 언제든 dm 이나 스레드에 멘션과 함께 요청을 해주시면 확인하고 설명드리겠습니다!
모든 것은 제 개인적인 생각과 의견이니 받아들일 부분은 받아들여주시면 좋고, 아니라면 자신만의 의견을 갖고 가시는 것도 좋습니다
첫 미션인데 정말 잘 해주셨고 고생 많으셨습니다

chxghee committed:
Fix: 리뷰어님 피드백 반영
의존성 주입 패턴으로 랜덤값 생
테스트 코드 수정

be-student commented on src/main/java/random/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,12 @@
+    private static final int RANDOM_NUMBER_BOUND = 10;
+
+    @Override
+    public int generate() {
```
이 부분 바꿔주신 것 확인했습니다!
고생 많으셨어요 🙇 
이제 딱 1개만 더 되면 훨씬 성능상 안정적인 코드가 될 것 같아요! 혹시 ThreadLocalRandom.current() 도 필드로 옮겨주실 수 있으신가요?
사실 ThreadLocalRandom.current() 라는 것도 함수다보니 내부적으로는 약간의 오버헤드가 있는 편입니다 따라서 이걸 그냥 필드 변수로 가지고 재사용 하면 됩니다!


<img width="765" alt="스크린샷 2025-03-17 오전 12 24 04" src="https://github.com/user-attachments/assets/9b502c8b-0d8d-4a0c-997e-3d313e734e20" />

chxghee replied:
반영했습니다~~~!!

be-student commented on src/main/java/domain/Racing.java:
```diff
@@ -0,0 +1,57 @@
+public class Racing {
+
+    private List<Car> cars = new ArrayList<>();
+    private List<Car> winners = new ArrayList<>();
```
제가 적어드린 것을 반영해주셨군요!
모든 코멘트에 항상 앞에는 이 말을 템플릿처럼 작성해두고 싶은데요
제 말은 언제나 정답이 아닐 수 있다는 것을 기억하면서 자기만의 방향을 찾아가시면 좋을 것 같습니다
저도 그냥 평범한 주니어니까요

이제 Racing 이라는 객체에 필드 변수에 winners 라는 필드가 생기게 되었는데, 이 필드는 창희님이 보시기에 여기에 있는 것이 자연스러운가요?
맥락상에서는 충분히 자연스러울 수 있는데, 이 클래스를 이제 잘못 사용할 수 있는 방법이 생기고 말았어요 😢 
new Racing(numberGenerator).getWinners()를 바로 호출하게 되면 이 부분은 의도하지 않은 동작이 되어버리고 만거죠
무조건 빈 리스트가 나올테니까요

그렇다면 어떻게 이 부분을 더 좋은 방향으로 바꿔볼 수 있을 것 같으신가요?
창희님이 생각하셨던 방향과 장단점에대해서도 설명해주세요

chxghee replied:
저도 코드를 작성할때 이 부분이 조금 걸렸었는데요, 이러한 문제 때문에 처음에는 필드로 우승자들을 저장하지 않으려고 했습니다.
처음 피드백을 받고 리스트로 저장해 놓는 방향으로 개선을 했는데,
그때는 아직 우승자가 정해지지 않았다면 빈 리스트를 반환하는게 맞는거 아닌가? 하는 생각으로 
그냥 get메서드에서 예외를 처리하지 않고 넘어가는 걸로 결론냈었습니다.

하지만 외부에서 get메서드를 호출할 때 리스트가 비어 있는지 확인을 하는 절차없이 값을 사용하게 된다면 의도와 달리 동작할 수 있기 때문에 getWinners를 호출할 때 현재 상태가 비어 있으면 IllegalStateException 예외를 던지도록 수정하였습니다

장점은 명확하게 예외상황에서 예외를 발생시켜 의도와 다른 동작을 차단하는 것이 장점 같고,
이 방식의  단점은 이 메서드를 호출하는 외부에서 예외를 처리하는 부분이 필요한 것이 단점인것 같습니다
(혹시 예외를 던지는 것 이외에 다른 해결 방법이 있을까요?)

그리고 다른 부분도 생각해 보니 레이싱이 시작되지도 않았는데 determineWinners() 를 호출하게 되면 이 부분도 큰  문제가 될것 같아서 private 로 메서드를 바꾸고 startRacing()메서드 안에서 determineWinners 를 호출하도록 바꾸어 보았습니다!

be-student commented on src/main/java/domain/Racing.java:
```diff
@@ -0,0 +1,57 @@
+    private List<Car> winners = new ArrayList<>();
+    private final NumberGenerator numberGenerator;
+
+    public Racing(NumberGenerator numberGenerator) {
```
이렇게 바꾸신 후 어떤 부분이 바뀐 것 같으신가요?
이렇게 하면 좋겠다고 말씀드린 것은 제 개인적인 의견이다보니 분명히 좋은 점도, 단점도 있을텐데요
써보니 어땠다 라고 하는 의견이 궁금해요

chxghee replied:
이렇게 바꾸니 외부에서(이 클래스를 생성하는 상위 클래스) 인터페이스 구현체를 갈아 끼워 랜덤 숫자 생성을 핸들링 할 수 있는게 큰 장점 같습니다 

이렇게 제어할 수 없는 값들을 최대한 컨트롤러같은 프로그램의 진입점에 위치시켜서 주입하도록 해 사용되는 메서드들을 최대한 테스트 가능하게 만드는 것이 최우선 인것 같고, 

인터페이스를 통해 선택하여 값을 주입하면 랜덤값에 대해 밀접하게 연관된 메서드 또한 테스트가 가능해 지네요
인터페이스로 생성되는 랜덤값을 선택하는 것은 저는 생각해 보지 못했는데 좋은 방법 같습니다!

be-student commented on README.md:
```diff
@@ -0,0 +1,40 @@
+ ✅ 자동차 경주 게임을 완료한 후 누가 우승했는지를 구할 수 있다. 우승자는 한 명 이상일 수 있다.
+
+
+### 💻 구현 전략
+1. Racing 객체에 레이싱에 참여하는 자동차들을 리스트로 저장할 필드 추가
+2. 주어진 n값 만큼 자동차를 생성하여 리스트에 저장
+3. 모든 자동차 각각에 대해 랜덤값을 할당하고 자동차를 이동시킬 메서드
+4. 주어진 횟수만큼 각 라운드를 진행할 메서드
+5. 우승자를 가려내는 메서드
```
이 구현 전략에 대해서 조금 더 자세하게 설명이 있으면 좋을 것 같아요
아마도 리드미를 체크리스트 형태로 사용한 것 같은데요 리드미는 다른 사람에게 설명해주기 위한 문서이다보니 이런 부분이 문서에 잘 드러나면 좋을 것 같아요!
지금의 경우에는 단순 설명보다는 무엇을 고민했는지, 그 결과가 어떤 것이었는지를 적어주시면 좋습니다

chxghee replied:
제가 나름 고민한 부분을 아래 새로운 항목으로 추가해서 적어보도록 할게요!

be-student commented on src/main/java/random/NumberGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package random;
```
여기서 random 이라는 것은 어떤 패키지명을 의미하고 있나요?
domain 과는 다른 기준인 것 같아서 궁금해요

chxghee replied:
이렇게 분리한 이유는 랜덤 숫자를 생성하는 것들을 한 패키지에 묶어 놓고 싶어서 였습니다.

하지만 이 숫자 생성하는 인터페이스 구현체는 레이싱 도메인에 특화되어 사용되고 있기 때문에 아래와 비슷한 이유로 이 도메인에 포함하는 것이 더욱 괜찮은 방법일까요?

be-student commented on src/main/java/random/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.concurrent.ThreadLocalRandom;
+
+public class RandomNumberGenerator implements NumberGenerator {
+    private static final int RANDOM_NUMBER_BOUND = 10;
```
이 부분은 예전에 아마 제가 받았었나 다른 사람이 받았거나 했던 리뷰였던 것으로 기억해요
RandomNumberGenerator 는 누군가가 생각하는 것처럼 진짜 random 을 지원하나요?
모르는사람이 RandomNumberGenerator 인데, 왜 10까지만 만들도록 만들어뒀지? 라는 의문을 가지고 이를 20으로 바꾸게 되면 분명히 사이드이펙트가 발생할 수 있을 것으로 보여요

만약 다른곳에서 사용하는 것을 원하지 않는다면 일반적인 이름인 RandomNumberGenerator 대신 도메인에 특화된 이름을 가지는 것이 좋을 것 같아요

chxghee replied:
다른곳에서 사용될 수 있는 여지를 줄이려면 우선 패키지 또한 도메인으로 위치하도록 바꾸고,
레이싱이라는 도메인에 특화된 이름을 갖도록 RacingNumberGenerator 로 수정하였습니다

생각하지 못한 부분이었는데 항상 내 코드를 다른사람과 같이 작성한다고 생각하면서 해야겠군요!

be-student reviewed (approved):
첫 리뷰였는데 어떠셨나요
리뷰에 적어둔 것처럼 평상시에 궁금한 것들이 있다면 언제든 여쭤볼 수 있는 그런 시간이 되었으면 좋았을 것 같은데 그런 시간이 된 것 같으신가요?

참고로 리뷰를 반영하거나 뭔가 질문사항이 있으면 코멘트에 추가로 달아주거나 이모지를 달아주는 것이 아 이 사람이 이걸 고려했구나 것을 리뷰해주는 쪽에서도 인지할 수 있다보니 서로에게 도움이 많이 되는 습관이에요!

질문 사항이 있다면 언제든 dm 이나 댓글로 남겨주세요

chxghee committed:
Fix: 리뷰어님 리뷰 반영2
getWinners()메서드 예외처리 추가
랜덤값 생성 클래스 이름 수정 RandomNumberGenerator -> RacingNumberGenerator

chxghee committed:
Docs: README.md 수정
